### ISSUE TITLE
Testing the model : `RuntimeWarning: All-NaN slice encountered`

### ISSUE THREAD

**Issue by Yassine-Fak** [ultralytics/yolov5#3331]
Hello, I am using 6 .tiff images to test my yolov5 model (model trained on jpg images, on one class using Transfer Learning).

While testing the model, using the following command:  

`python test.py --weights models/best.pt --data data/data.yml --img-size 416 --task test --conf 0.8 --batch-size 2` I got the following Warning: 

```
Speed: 14.1/0.2/14.3 ms inference/NMS/total per 416x416 image at batch-size 2
/home/anaconda3/envs/yolov5/lib/python3.8/site-packages/seaborn/matrix.py:194: RuntimeWarning: All-NaN slice encountered
  vmin = np.nanmin(calc_data)
/home/anaconda3/envs/yolov5/lib/python3.8/site-packages/seaborn/matrix.py:199: RuntimeWarning: All-NaN slice encountered
  vmax = np.nanmax(calc_data)
Results saved to runs/test
```
Can you help me please to understand this Warning? Is it a critical warning to get an accurate test of the model?

Thank you in advance for your help

**Comment by Yassine-Fak** [comment]
Hey, I found the solution: my model still has poor performances, so I have to test using a lower threshold (in order to avoid getting Nan values)  

Have a nice day :) And many thanks again for this amazing repo

**Comment by ouyuanzheng** [comment]
How did you solve the problem "RuntimeWarning: All-NaN slice encountered"? I ran into this problem and it can't be solved, you have any advice？

**Comment by ouyuanzheng** [comment]
the following Warning:
PRECISION OPTIMIZATION FAILED, skipping...
/mnt/d/code-3Dmap2/iarpa/functions.py:363: RuntimeWarning: All-NaN slice encountered
  min_np = np.nanmin(height_np)
/mnt/d/code-3Dmap2/iarpa/functions.py:366: RuntimeWarning: All-NaN slice encountered
  max_height = np.nanmax(height_np)
(3, 2.577479839324951)
(1, 1.5050010681152344)
(2, 2.577479839324951)
('Total time:', 4.0848870277404785)

**Comment by RoeschFlo** [comment]
Hello Yassine-Fak,
can you tell me which parameter you have chanced? :D
txh

**Comment by glenn-jocher** [comment]
@RoeschFlo hello Yassine-Fak,

Based on the warning message you provided, it seems that you encountered an issue with a "RuntimeWarning: All-NaN slice encountered". This warning typically occurs when there are NaN (Not a Number) values in your data.

To solve this issue, you can try the following steps:

1. Check your input data: Ensure that your input data does not contain any NaN values. NaN values can sometimes cause issues during calculations, leading to this warning.

2. Lower the threshold: If your model is still yielding NaN values, you can try reducing the confidence threshold (--conf parameter) during inference. Lowering the threshold allows the model to detect more objects, potentially avoiding NaN values.

Additionally, it would be helpful if you can provide more details about the exact steps and parameters you used, as well as any changes you made, so that we can further investigate the issue and provide more specific guidance.

Best regards,
Glenn Jocher

**Comment by RoeschFlo** [comment]
i found out, i named the labels in the val file wrong.  thx for your help.

**Comment by glenn-jocher** [comment]
@RoeschFlo great to hear that you found the solution to the issue! It's not uncommon to overlook small details like naming labels, so don't worry about it. If you have any further questions or encounter any other issues in the future, feel free to ask for help. Keep up the good work!

**Comment by LucaGare** [comment]
Hello, I get the same warning, but during training
(and using yolo_v7, I don't know if you could help me with that)

The label files looks fine to me

**Comment by glenn-jocher** [comment]
@LucaGare hi there,

I'm sorry to hear that you're encountering a warning during training with YOLOv5 and yolo_v7. It would be helpful if you could provide more information about the specific warning you're seeing and any other relevant details such as your dataset, configuration files, and training command.

In addition, please double-check your label files to ensure they are properly formatted and correspond correctly to your training images. Any inconsistencies or errors in the label files could potentially lead to warnings or other issues during training.

If you can provide more specific details and any error messages you are seeing, I'll be happy to help you further troubleshoot the problem.

Regards,
Glenn Jocher

**Comment by LucaGare** [comment]
Hi @glenn-jocher,

**Regarding the dataset**:
It is made of ~1600 images (~1200 train, ~200 valid, ~200 test). 
I have 1 single class that I'm trying to detect.

**Regarding the configuration**:
- dataset yaml file. In data/ I've created a yaml file called 'events.yaml'  
![image](https://github.com/ultralytics/yolov5/assets/79198549/10f89dbb-84c6-474b-a454-1536b27e8611)
- model cfg yaml file. in cfg/training/ I've crated a copy of 'yolov7-tiny.yaml', I've named it 'yolov7-tiny_custom.yaml', the only difference to the original file is the value of 'nc' (number of classes) at line 2. In my case nc is set to 1

**My training command (I'm doing the training using CPU, not GPU)**:
python train.py --epochs 30 --workers 4 --batch-size 8 --data data/events.yaml --img 320 320 --cfg cfg/training/yolov7-tiny_custom.yaml --weights 'yolov7-tiny.pt' --name my_custom_training --hyp data/hyp.scratch.tiny.yaml

**Issue I get during the training**:
After some epochs I start getting 'nan' for 'box', 'obj' and 'total'
![image](https://github.com/ultralytics/yolov5/assets/79198549/6d139f00-e242-4d17-90e7-1cca77275c13)
(with higher values of workers, I get the 'nan' values earlies, even from the first epoch)

**Regarding the specific warning I'm getting at the end of the training proccess**:
/home/miniconda3/envs/yolov7/lib/python3.8/site-packages/seaborn/matrix.py:202: RuntimeWarning: All-NaN slice encountered
  vmin = np.nanmin(calc_data)
/home/miniconda3/envs/yolov7/lib/python3.8/site-packages/seaborn/matrix.py:207: RuntimeWarning: All-NaN slice encountered
  vmax = np.nanmax(calc_data)
30 epochs completed in 3.028 hours.

**Regarding the label files**:
I performed the following checks.
- each line in each label file starts with '0'. (since I only have 1 class)
- each line in each label file contains values between 0 and 1.


Let me know if you need any other info.

**Comment by glenn-jocher** [comment]
@LucaGare hi, 

Regarding the issue you're experiencing during training, it seems that you're getting NaN values for 'box', 'obj', and 'total' after some epochs. This can happen if the network fails to converge or encounters numerical instability.

To further investigate this issue, could you provide some additional details or clarifications?

1. Are you using any custom loss functions or modifications to the core YOLOv5 algorithm?

2. What is the range of the loss values before encountering NaN? Are they extremely large or small?

3. Have you tried adjusting the learning rate (--lr) or other hyperparameters to see if it helps stabilize the training?

4. Could you share the contents of your custom .yaml file (data/events.yaml) for the dataset? Are there any anomalies or inconsistencies in the data or label paths?

By providing this information, we can better understand and troubleshoot the issue you're facing during training.

Thanks in advance for your assistance.

-Glenn

**Comment by LucaGare** [comment]
@glenn-jocher Hi and thanks for the support.

1. I am not using any custom loss functions or modifications. However, as pointed out earlier, I am not using YOLOv5, but rather YOLOv7 (hoping it is similar enough regarding the issue I'm facing). Particularly the YOLOv7 tiny model.

2. Where can I inspect the loss values?
In the meantime, here is an extract of the 'results.txt" file that is generated during the training:
![image](https://github.com/ultralytics/yolov5/assets/79198549/45d8027b-45a6-4c08-a2ed-1f7128d589b5)

3. No, I left the default values for the hyperparameters. What would you suggest to try to modify?

4. My custom .yaml file is exactly what I posted in my previous message, i.e.
![image](https://github.com/ultralytics/yolov5/assets/79198549/3200f39f-27b8-4702-bcff-e583ef53707a)
I see no inconsistencies in the data or label paths. The images are in the path specified in the file and the labels are in ../storm_dataset/labels/train (and similar for valid and test)

Thanks a lot.
Luca

**Comment by glenn-jocher** [comment]
@LucaGare

Thanks for providing additional information. In YOLOv7, the loss values can be inspected from the live training progress bar or directly from the training log. From the results you provided, it seems that both the bounding box loss (box) and objectness loss (obj) are reaching "nan" values, which indicates that the model is encountering numerical instability during training.

Given that you're using the default hyperparameters, let's try adjusting the learning rate (--lr) to a lower value to see if it helps stabilize the training. You can consider using a smaller learning rate such as 0.001 or 0.0001. Additionally, you may want to experiment with different values for the momentum (--momentum) and weight decay (--weight-decay) to see if they have an impact on the training stability.

Lastly, ensure that your label files are correctly formatted and that the image and label paths specified in your custom .yaml file are accurate.

I hope these suggestions help in addressing the issue you're facing. Let me know if you have any further questions or if you need assistance with anything else.

Best regards,  
Glenn

**Comment by RoeschFlo** [comment]
Hi :D
Is there a way how you can run:" python3 test.py" on a test dataset that contains only background? I would like to test the robustness of the model against false positives with different thresholds. It would be nice if i had out of the box metrics. is this somehow possible with the default code from the repo? because when i use empty label txt files i get the info that nothing was detected. 
I am asking, because i asumed, that the metrics would somehow work in this usecase too.
But i think i will just count the number of outputs in detect.py and thats the number of false positives.


Flo

**Comment by glenn-jocher** [comment]
@RoeschFlo

You can indeed test the YOLOv5 model on a test dataset consisting only of background images to evaluate its robustness against false positives. To simulate this scenario, you can create a test dataset containing background images and use empty label files alongside those images. This will allow you to observe the model's behavior when detecting objects in the absence of true positives.

While the current testing script might not provide metrics specifically for the false positive scenario, you can customize the "test.py" script to count the number of detections on the background images. By inspecting the output, you can compute the number of false positives and adjust the confidence threshold to observe its impact on the false positive rate.

If you need further guidance on modifying the "test.py" script or have any additional questions, feel free to ask for assistance. I'm here to help.

Cheers,  
Glenn

**Comment by RoeschFlo** [comment]
@glenn-jocher 

Hi, thanks for the fast answer. I opened a new question concerning my difficulties. 
https://github.com/WongKinYiu/yolov7/issues/1916

**Comment by glenn-jocher** [comment]
@RoeschFlo great to hear that you've opened a new question. I hope you receive valuable insights and solutions to address your difficulties! If you have any further questions or need additional assistance, feel free to reach out to the YOLOv5 community. 

Best of luck with your query on the YOLOv7 repository!

Glenn
